src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/HtmlHelper.cs

```diff
@@ -748,7 +749,18 @@ protected virtual string GenerateDisplayText(ModelExplorer modelExplorer)
             if (tagBuilder != null)
             {
                 tagBuilder.TagRenderMode = TagRenderMode.StartTag;
-                tagBuilder.WriteTo(ViewContext.Writer, _htmlEncoder);
+
+                // As a perf optimization, we can buffer this output rather than writing it
+                // out character by character.
+                var htmlWriter = ViewContext.Writer as HtmlTextWriter;
+                if (htmlWriter == null)
+                {
+                    tagBuilder.WriteTo(ViewContext.Writer, _htmlEncoder);
+                }
+                else
+                {
+                    htmlWriter.Write(tagBuilder);
+                }
```

Comment:
Move the special case down into `TagBuilder.WriteTo(...)` and pick up calls to that method from `RazorPage` for free. At least, similar code is necessary in `GenerateRouteForm()`.